docs/topics/functions.md

```diff
@@ -40,72 +65,168 @@ fun powerOf(
+like the `len` parameter in this example:
 
 ```kotlin
-fun foo(
-    bar: Int = 0,
-    baz: Int,
+fun read(
+    b: ByteArray,
+    off: Int = 0,
+    len: Int = b.size,
 ) { /*...*/ }
-
-foo(baz = 1) // The default value bar = 0 is used
 ```
 
-If the last parameter after all parameters with default values has a functional type,
-then you can pass the corresponding [lambda](lambdas.md#lambda-expression-syntax) argument either as a named argument or [outside the parentheses](lambdas.md#passing-trailing-lambdas):
+Parameters that refer to the values of other parameters must be declared later in the order.
+In this example, `len` must be declared after `b`.
+
+In general, you can assign any expression as the default value of a parameter.
+However, default values are only evaluated when the function is called **without** the corresponding parameter
+and a default value needs to be assigned.
+For example, this function prints out a line only when it is called without the `print` parameter:
 
 ```kotlin
-fun foo(
-    bar: Int = 0,
-    baz: Int = 1,
-    qux: () -> Unit,
-) { /*...*/ }
+fun main() {
+//sampleStart
+fun read(
+    b: Int,
+    print: Unit? = println("No argument passed for 'print'")
+) { println(b) }
+
+// Prints "No argument passed for 'print'", then "1"
+read (1)
+// Prints only "1"
+read (1, null)
+//sampleEnd
+}
+```
+{kotlin-runnable="true" kotlin-min-compiler-version="1.3" id="non-constant-default"}
+
+If the last parameter in a function declaration has a functional type,
+you can pass the corresponding [lambda](lambdas.md#lambda-expression-syntax) argument either as a named argument or [outside the parentheses](lambdas.md#passing-trailing-lambdas):
 
-foo(1) { println("hello") }     // Uses the default value baz = 1
-foo(qux = { println("hello") }) // Uses both default values bar = 0 and baz = 1
-foo { println("hello") }        // Uses both default values bar = 0 and baz = 1
+```kotlin
+fun main() {
+//sampleStart
+fun log(
+    level: Int = 0,
+    code:  Int = 1,
+    action: () -> Unit,
+) { println (level)
+    println (code)
+    action() }
+
+// Passes 1 for 'level' and uses the default value 1 for 'code'
+log(1) { println("Connection established") }
+
+// Uses both default values, 0 for 'level' and 1 for 'code'
+log(action = { println("Connection established") })
+
+// Equivalent to the previous call, uses both default values
+log { println("Connection established") }
+//sampleEnd   
+}
 ```
+{kotlin-runnable="true" kotlin-min-compiler-version="1.3" id="lambda-outside-parentheses"}
 
 ### Named arguments
 
-You can name one or more of a function's arguments when calling it. This can be helpful when a function has many
-arguments and it's difficult to associate a value with an argument, especially if it's a boolean or `null` value.
+You can name one or more of a function's arguments when calling it.
+This can be helpful when a function call has many argument, and it's difficult to associate a value with an argument,
```

Comment:
```suggestion
This can be helpful when a function call has many arguments. In this case, it can be difficult to associate a value with an argument,
```